Project refactoring

Sam09040 opened the pull request:
## added:
- instaq file structure;
- .mocharc.json for mocha config;
- paths;

Sam09040 committed:
Project refactoring

## added:
- instaq file structure;
- .mocharc.json for mocha config;
- paths;

racsandoval commented on src/api/graphql/resolvers.ts:
```diff
@@ -1,29 +1,17 @@
-    ) => usersQuery(skip, limit, context.token),
-    user: async (_: unknown, { id }: { id: number }, context: ContextType): Promise<User> => userQuery(id, context.token),
-    address: async (_: unknown, { userId }: { userId: number }) => addressQuery(userId),
+    user: userResolver.Query.user,
+    users: userResolver.Query.users,
+    address: addressResolver.Query.address,
```
```suggestion
    ...userResolver.Query
```

racsandoval commented on src/domain/address/address.use-case.ts:
```diff
@@ -0,0 +1,26 @@
+const userDatasource = new UserDbDataSource();
+
+export async function addressUseCase(userId: number): Promise<AddressModel[]> {
+  if (!userId) {
+    throw new InvalidDataError('No user provided', {
+      field: 'userId',
+      reason: 'A user ID is required to fetch the address.',
+    });
+  }
```
geralmente colocamos as checagens dos inputs que vem nas chamadas na própria camada de API, os use-cases sempre devem receber os dados corretamente

racsandoval commented on src/domain/model/address.model.ts:
```diff
@@ -0,0 +1,24 @@
+  state: string;
+}
+
+export interface AddressInputModel {
+  data: {
+    userId: number;
+    cep: string;
+    street: string;
+    streetNumber: string;
+    complement?: string | null;
+    neighborhood: string;
+    city: string;
+    state: string;
+  };
+}
```
pq o input fica dentro de um objeto chamada data? não daria pra interface ser os campos diretamente?
```suggestion
export interface AddressInputModel {
    userId: number;
    cep: string;
    street: string;
    streetNumber: string;
    complement?: string | null;
    neighborhood: string;
    city: string;
    state: string;
}
```

racsandoval commented on src/test/utils/get-token.util.ts:
```diff
@@ -0,0 +1,33 @@
+import { requestMaker } from '@test/request-maker';
+import { LoginInputModel } from '@domain/model';
+
+export const getToken = async (): Promise<string | undefined> => {
```
essa função faz uma requisição pra chamada de login e é utilizada nos outros testes para autenticar o usuário. Não é uma boa prática usar uma chamada dentro do teste de outra, já que se a chamada de login quebrar todos os outros testes quebram, portanto todos ficam dependentes da chamada de login

Sam09040 committed:
small fix

- Removed data from the input models;
- Changed getToken function to only create and return token;

racsandoval commented on src/api/graphql/resolvers.ts:
```diff
@@ -1,29 +1,15 @@
-    user: async (_: unknown, { id }: { id: number }, context: ContextType): Promise<User> => userQuery(id, context.token),
-    address: async (_: unknown, { userId }: { userId: number }) => addressQuery(userId),
+    ...userResolver.Query,
+    address: addressResolver.Query.address,
```
poderia fazer a mesma coisa com as queries de address, mesmo tendo apenas uma, pois não vai precisar alterar esse arquivo quando forem adicionadas mais chamadas de endereço

racsandoval commented on src/domain/user/create-user.use-case.ts:
```diff
@@ -0,0 +1,56 @@
+const bcryptService = new BcryptService();
+
+export async function createUserUseCase(input: UserInputModel, token: string | undefined): Promise<UserModel> {
+  if (!token) {
+    throw new UnauthorizedError('Token is required for this operation!', {
+      field: 'authorization',
+      reason: 'A valid token must be provided.',
+    });
+  }
+
+  jwtService.verify(token);
+  const { name, email, password, birthDate } = input;
+  if (!name || !email || !password || !birthDate) {
+    throw new InvalidDataError('Invalid input!', {
+      field: 'data',
+      reason: 'All fields are required!',
+    });
+  }
```
essas verificações poderia estar na camada de API, já que elas são relacionadas a campos da chamada. o domínio não deveria conhecer o token e nem ter que validar os inputs da camada de api

Sam09040 committed:
fix: verifications

- got the input and token verification off of the use cases and put it on resolvers;
